File: epistimi-frontend/src/dto/course.ts
import { StudentResponse } from './student';
import { TeacherResponse } from './teacher';

export interface CourseResponse {
  id: string;
  code: CourseCode;
  schoolYear: string;
  classTeacher: TeacherResponse;
  students: StudentResponse[];
  subjects: CourseSubjectResponse[];
  schoolYearBegin: Date;
  schoolYearSemesterEnd: Date;
  schoolYearEnd: Date;
  profile?: string;
  profession?: string;
  specialization?: string;
}

export interface CourseCode {
  number: string;
  letter: string;
}

export interface CourseSubjectResponse {
  id: string;
  teacher: CourseSubjectTeacherResponse;
  name: string;
}

export interface CourseSubjectTeacherResponse {
  id: string;
  academicTitle: string | null;
  firstName: string;
  lastName: string;
  username: string;
}

export interface CoursesResponse {
  courses: CourseResponse[];
}

export interface CourseFormData {
  codeNumber?: number;
  codeLetter: string;
  schoolYearBegin?: Date;
  schoolYearSemesterEnd?: Date;
  schoolYearEnd?: Date;
  classTeacherId: string;
  profile?: string;
  profession?: string;
  specialization?: string;
}

export interface CourseCreateRequest {
  codeNumber?: number;
  codeLetter: string;
  schoolYearBegin?: string;
  schoolYearSemesterEnd?: string;
  schoolYearEnd?: string;
  classTeacherId: string;
  profile?: string;
  profession?: string;
  specialization?: string;
}

File: epistimi-backend/src/main/kotlin/pl/edu/wat/wcy/epistimi/course/domain/Course.kt
package pl.edu.wat.wcy.epistimi.course.domain

import org.hibernate.annotations.GenericGenerator
import pl.edu.wat.wcy.epistimi.organization.domain.Organization
import pl.edu.wat.wcy.epistimi.student.domain.Student
import pl.edu.wat.wcy.epistimi.subject.domain.Subject
import pl.edu.wat.wcy.epistimi.teacher.domain.Teacher
import java.time.LocalDate
import java.util.UUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "courses")
class Course(
    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @Column(name = "id", length = 36, nullable = false, updatable = false)
    val id: CourseId? = null,

    @ManyToOne
    val organization: Organization,

    @Column(name = "code_number", nullable = false)
    val codeNumber: Int,

    @Column(name = "code_letter", nullable = false)
    val codeLetter: String,

    @ManyToOne(optional = false)
    val classTeacher: Teacher,

    @OneToMany
    @JoinColumn(name = "course_id")
    val students: Set<Student>,

    @OneToMany
    @JoinColumn(name = "course_id")
    val subjects: Set<Subject>,

    @Column(name = "school_year_begin", nullable = false)
    val schoolYearBegin: LocalDate,

    @Column(name = "school_year_semester_end", nullable = false)
    val schoolYearSemesterEnd: LocalDate,

    @Column(name = "school_year_end", nullable = false)
    val schoolYearEnd: LocalDate,

    @Column(name = "profile")
    val profile: String?,

    @Column(name = "profession")
    val profession: String?,

    @Column(name = "specialization")
    val specialization: String?,
) {
    val code
        get() = "$codeNumber$codeLetter"

    val schoolYear
        get() = "${schoolYearBegin.year}/${schoolYearEnd.year}"
}

@JvmInline
value class CourseId(
    val value: UUID,
)
